### PR TITLE
Add Support for Deploying SDDC Manager appliance with Pod

### PR DESCRIPTION
**Pull request by rexit1982**
Add Support for Deploying SDDC Manager appliance with Pod, can later be used for VCF Import to get full stack management.
Next Steps would probably be automating copying the import tool to SDDCMgr and vcsa to save time.

### CODE REVIEW

File: playbooks/DeploySDDCMgr.yml [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        allow_duplicates: false
+        # For vCenter deployment, VM Folder path must be the FULL path: /<DataCenter>/vm/<ParentFolder>/<ChildFolder>
+        folder: "{% if Target.Deployment == 'vCenter' %}/{{ Target.DataCenter }}/vm/{{ Target.VMFolder }}\
+                 {% else %}{{ Target.VMFolder }}{% endif %}"
+        datacenter: "{{ Target.DataCenter }}"
+        cluster: "{{ Target.Cluster }}"
+        name: "{{ Nested_SDDCMgr.Component.Appliance.VMName }}"
+        ovf: "{{ Deploy.Software.SDDCMgr.Directory }}/{{ Deploy.Software.SDDCMgr.File }}"
+        networks:
+          "Management": "{{ Target.PortGroup.Management.Name }}"
```
Comment (rutgerblom) [inline playbooks/DeploySDDCMgr.yml]: I had to change this from "Management" to "Network 1" in order for the OVF deployment to start. Is it the same on your side?

File: playbooks/DeploySDDCMgr.yml [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+          ROOT_PASSWORD: "{{ Nested_SDDCMgr.Credential.root.Password }}"
+          VCF_PASSWORD: "{{ Nested_SDDCMgr.Credential.vcf.Password }}"
+          BASIC_AUTH_PASSWORD: "{{ Nested_SDDCMgr.Credential.admin.Password }}"
+          BACKUP_PASSWORD: "{{ Nested_SDDCMgr.Credential.backup.Password }}"
+          LOCAL_USER_PASSWORD: "{{ Nested_SDDCMgr.Credential.local.Password }}"
+          FIPS_ENABLE: "{{ Nested_SDDCMgr.Component.Appliance.FIPSEnable }}"
+          vami.hostname: "{{ Nested_SDDCMgr.Component.Appliance.FQDN }}"
+          guestinfo.ntp: "{{ Common.NTP.Server1.IPv4 }}"
+          vami.ip0.SDDC-Manager: "{{ Nested_SDDCMgr.Component.Appliance.Address.IPv4.Address }}"
+          vami.netmask0.SDDC-Manager:  "{{ (Nested_SDDCMgr.Component.Appliance.Address.IPv4.Address + '/' + Nested_SDDCMgr.Component.Appliance.Address.IPv4.Prefix) | ansible.utils.ipaddr('netmask') }}"
```
Comment (rutgerblom) [inline playbooks/DeploySDDCMgr.yml]: Ansible lint complains about the extra space after colon.

File: playbooks/DeploySDDCMgr.yml [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                    SDDCMgr root User: {{ Nested_SDDCMgr.Credential.root.Name }}
+                    SDDCMgr root Password: {{ Nested_SDDCMgr.Credential.root.Password }}
+                    SDDCMgr admin User: {{ Nested_SDDCMgr.Credential.admin.Name }}
+                    SDDCMgr admin Password: {{ Nested_SDDCMgr.Credential.admin.Password }}
+                    SDDCMgr local User: {{ Nested_SDDCMgr.Credential.local.Name }}
+                    SDDCMgr local Password: {{ Nested_SDDCMgr.Credential.local.Password }}
+                    SDDCMgr backup User: {{ Nested_SDDCMgr.Credential.backup.Name }}
+                    SDDCMgr backup Password: {{ Nested_SDDCMgr.Credential.backup.Password }}
+                    SDDCMgr vcf User: {{ Nested_SDDCMgr.Credential.vcf.Name }}
+                    SDDCMgr vcf Password: {{ Nested_SDDCMgr.Credential.vcf.Password }}
```
Comment (rutgerblom) [inline playbooks/DeploySDDCMgr.yml]: Ansible lint complains about the trailing space.

File: playbooks/DeploySDDCMgr.yml (hunk 1 above)
Comment (rexit1982) [reply playbooks/DeploySDDCMgr.yml]: I did not and I ran a couple test runs before creating the PR.  looking over the ovf file I don't know why it ever worked like this.  File updated.

File: playbooks/DeploySDDCMgr.yml (hunk 2 above)
Comment (rexit1982) [reply playbooks/DeploySDDCMgr.yml]: Had to l;earn what ansible lint was... neat tool, alerts addressed.

File: playbooks/DeploySDDCMgr.yml (hunk 3 above)
Comment (rexit1982) [reply playbooks/DeploySDDCMgr.yml]: Had to l;earn what ansible lint was... neat tool, alerts addressed.

**Review by luischanu — COMMENTED** [review]
@rexit1982 feedback on two items:

1. Type-O in comment of ```DeploySDDCMgr.yml```.  ```SDDG``` should be ```SDDC```.  Not a big deal, but since I'm entering this comment, thought I'd mention it.

2. I can foresee where someone might want to deploy SDDC Manager in a deployment regardless of the DeployESXiOnly setting.   If they set it to false, then all other products will deploy in the Pod, which can then be imported.  If they set it to true, then they want to use SDDC Manager to deploy all the products.  So, not sure you need the ```Deploy.Setting.DeployESXiOnly == false``` check on the 'when' statement for those reasons.

@rutgerblom , do you agree?

Luis

File: CHANGELOG.md [hunk 4]
```diff
@@ -2732,4 +2732,16 @@
...
+  
+## Dev-v8.0.0 10-OCTOBER-2024
+
+### Added by Aaron Ellis
+  - Added support for deploying an SDDC Manager appliance during POD Build. Can be used for VCF Import, possibly automated later. 
+	- Added SDDC Manager v5.2 to the ```software_sample.yml``` file
+	- Added Nested_SDDC section to the ```config_sample.yml``` file
+	- Updated ```templates/Pod_Config.j2```
+	- Updated ```CreatePodConfig.yml```
+	- Added ```playbooks/DeploySDDCMgr.yml```
```
Comment (rutgerblom) [inline CHANGELOG.md:2745]: A small detail but I agree with Luis feedback. Let's go for playbooks/DeploySDDCManager.yml for improved readability.

File: Deploy.yml [hunk 5]
```diff
@@ -248,6 +248,12 @@
     - Deploy.Product.Replication.Deploy
     - Deploy.Setting.DeployESXiOnly == false
 
+# Deploy SDDG Manager (Nested)
+- ansible.builtin.import_playbook: playbooks/DeploySDDCMgr.yml
```
Comment (rutgerblom) [inline Deploy.yml:252]: DeploySDDCManager.yml for improved readability.

File: Deploy.yml [hunk 6]
```diff
@@ -248,6 +248,12 @@
...
+  when: 
+    - Deploy.Product.SDDCMgr.Deploy
```
Comment (rutgerblom) [inline Deploy.yml:254]: DeploySDDCManager.yml for improved readability.

File: Deploy.yml [hunk 7]
```diff
@@ -248,6 +248,12 @@
...
+    - Deploy.Setting.DeployESXiOnly == false
```
Comment (rutgerblom) [inline Deploy.yml:255]: This conditional can be removed as per Luis comment.

File: README.md [hunk 8]
```diff
@@ -265,6 +265,7 @@ When a Pod is deployed, various components are deployed as part of that Pod.  Ea
 | 23 | vRNI Platform | vRealize Network Insight Platform Appliance | Pod-240-vRNI | No |
 | 24 | vRNI Collector | vRealize Network Insight Collector Node | Pod-240-vRNI-Collector | No |
 | 25 | Replication Appliance | vSphere Replication Appliance | Pod-240-Replication | No |
+| 26 | SDDC Manager Appliance | SDDC Manager Appliance | Pod-240-SDDCMgr | No |
```
Comment (rutgerblom) [inline README.md:268]: Pod-240-SDDCManager improved readability.

File: config_sample.yml [hunk 9]
```diff
@@ -74,6 +74,8 @@ Deploy:
       Deploy: false                                       # ALB - Deploy NSX Advanced Load Balancer as part of the Pod deployment.
     Replication:
       Deploy: false                                       # Replication - Deploy a vSphere Replication Appliance as part of the Pod deployment.
+    SDDCMgr:
```
Comment (rutgerblom) [inline config_sample.yml:77]: SDDCManager: for improved readability.

File: config_sample.yml [hunk 10]
```diff
@@ -160,6 +162,14 @@ Deploy:
       Patch:   false
       Config:  "{{ SiteCode }}-Replication"
       Installer: "{{ SiteCode }}-Replication"
+    SDDCMgr:                                          # Label used for selection of which template to use during deployment (templates.yml)
```
Comment (rutgerblom) [inline config_sample.yml:165]: SDDCManager for improved readability.

File: config_sample.yml [hunk 11]
```diff
@@ -160,6 +162,14 @@ Deploy:
...
+      Vendor:  VMware                                     # Indicates software vendor  you want to deploy from the software repository (software.yml)
+      Product: SDDCMgr                                    # Indicates software product you want to deploy from the software repository (software.yml)
```
Comment (rutgerblom) [inline config_sample.yml:167]: SDDCManager for improved readability.

File: config_sample.yml [hunk 12]
```diff
@@ -160,6 +162,14 @@ Deploy:
...
+      Version: '5.2'                                      # Indicates which software version you want to deploy from the software repository (software.yml)
+      Patch:   false
+      Config:  "{{ SiteCode }}-SDDCMgr"
```
Comment (rutgerblom) [inline config_sample.yml:170]: "{{ SiteCode }}-SDDCManager" for improved readability.

File: config_sample.yml [hunk 13]
```diff
@@ -160,6 +162,14 @@ Deploy:
...
+      Installer: "{{ SiteCode }}-SDDCMgr"
```
Comment (rutgerblom) [inline config_sample.yml:171]: "{{ SiteCode }}-SDDCManager" for improved readability.

File: config_sample.yml [hunk 14]
```diff
@@ -811,6 +821,46 @@ Nested_Replication:
           Prefix:  "{{ Net.Management.IPv6.Prefix }}"
 
 
+##
+## ███    ██ ███████ ███████ ████████ ███████ ██████          ███████ ██████  ██████   ██████ ███    ███  ██████  ██████  
+## ████   ██ ██      ██         ██    ██      ██   ██         ██      ██   ██ ██   ██ ██      ████  ████ ██       ██   ██ 
+## ██ ██  ██ █████   ███████    ██    █████   ██   ██         ███████ ██   ██ ██   ██ ██      ██ ████ ██ ██   ███ ██████
```
Comment (rutgerblom) [inline config_sample.yml:827]: NESTED SDDCMANAGER for improved readability.

File: config_sample.yml [hunk 15]
```diff
@@ -811,6 +821,46 @@ Nested_Replication:
...
 
+##
+## ███    ██ ███████ ███████ ████████ ███████ ██████          ███████ ██████  ██████   ██████ ███    ███  ██████  ██████  
+## ████   ██ ██      ██         ██    ██      ██   ██         ██      ██   ██ ██   ██ ██      ████  ████ ██       ██   ██ 
+## ██ ██  ██ █████   ███████    ██    █████   ██   ██         ███████ ██   ██ ██   ██ ██      ██ ████ ██ ██   ███ ██████  
+## ██  ██ ██ ██           ██    ██    ██      ██   ██              ██ ██   ██ ██   ██ ██      ██  ██  ██ ██    ██ ██   ██ 
+## ██   ████ ███████ ███████    ██    ███████ ██████  ███████ ███████ ██████  ██████   ██████ ██      ██  ██████  ██   ██ 
+##                                                                                                                       
+
+Nested_SDDCMgr:
```
Comment (rutgerblom) [inline config_sample.yml:832]: Nested_SDDCManager: for improved readability.

File: config_sample.yml [hunk 16]
```diff
@@ -811,6 +821,46 @@ Nested_Replication:
...
+      Password: "{{ Common.Password.Nested }}{{ Common.Password.Nested }}"
+    backup:
+      Name: backup
+      Password: "{{ Common.Password.Nested }}{{ Common.Password.Nested }}"
+    vcf:
+      Name: vcf
+      Password: "{{ Common.Password.Nested }}{{ Common.Password.Nested }}"
+  Component:
+    Appliance:
+      VMName: "{{ SiteCode }}-SDDCMgr"
```
Comment (rutgerblom) [inline config_sample.yml:851]: "{{ SiteCode }}-SDDCManager" for improved readability.

File: playbooks/CreatePodConfig.yml [hunk 17]
```diff
@@ -22,6 +22,7 @@
                                   _vRLIv{% if Deploy.Product.vRLI.Deploy == false %}NONE{% else %}{{ Deploy.Software.vRLI.Version }}{% endif %}\
                                   _ALBv{% if Deploy.Product.ALB.Deploy == false %}NONE{% else %}{{ Deploy.Software.ALB.Version }}{% endif %}\
                                   _REPLv{% if Deploy.Product.Replication.Deploy == false %}NONE{% else %}{{ Deploy.Software.Replication.Version }}{% endif %}\
+                                  _SDDCMv{% if Deploy.Product.SDDCMgr.Deploy == false %}NONE{% else %}{{ Deploy.Software.SDDCMgr.Version }}{% endif %}\
```
Comment (rutgerblom) [inline playbooks/CreatePodConfig.yml:25]: {{ Deploy.Software.SDDCManager.Version }} for improved readability.

File: playbooks/CreatePodConfig.yml [hunk 18]
```diff
@@ -30,6 +31,7 @@
                       _vRLIv{% if Deploy.Product.vRLI.Deploy == false %}NONE{% else %}{{ Deploy.Software.vRLI.Version }}{% endif %}\
                       _ALBv{% if Deploy.Product.ALB.Deploy == false %}NONE{% else %}{{ Deploy.Software.ALB.Version }}{% endif %}\
                       _REPLv{% if Deploy.Product.Replication.Deploy == false %}NONE{% else %}{{ Deploy.Software.Replication.Version }}{% endif %}\
+                      _SDDCMv{% if Deploy.Product.SDDCMgr.Deploy == false %}NONE{% else %}{{ Deploy.Software.SDDCMgr.Version }}{% endif %}\
```
Comment (rutgerblom) [inline playbooks/CreatePodConfig.yml:34]: Change to SDDCManager for improved readability.

File: playbooks/DeploySDDCMgr.yml [hunk 19]
```diff
@@ -0,0 +1,182 @@
+##
+##    Project: SDDC.Lab
+##    Authors: Luis Chanu & Rutger Blom
+##   Filename: playbooks/DeploySDDCMgr.yml
+##
+---
+- name: DeploySDDCMgr.yml
```
Comment (rutgerblom) [inline playbooks/DeploySDDCMgr.yml:7]: DeploySDDCManager.yml

File: playbooks/DeploySDDCMgr.yml [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+  hosts: localhost
+  tasks:
+    - name: DeploySDDCMgr_Playbook
```
Comment (rutgerblom) [inline playbooks/DeploySDDCMgr.yml:10]: DeploySDDCManager_Playbook

File: playbooks/DeploySDDCMgr.yml [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+
+                                          Target.FQDN: {{ Target.FQDN }}
+                                          Target.User: {{ Target.User }}
+                                      Target.Password: {{ Target.Password }}
+                                       Target.Cluster: {{ Target.Cluster }}
+                                    Target.DataCenter: {{ Target.DataCenter }}
+                                     Target.Datastore: {{ Target.Datastore }}
+                                      Target.VMFolder: {{ Target.VMFolder }}
+
+                                          SDDCMgr Version: {{ Deploy.Software.SDDCMgr.Version }}
```
Comment (rutgerblom) [inline playbooks/DeploySDDCMgr.yml:64]: Replace with SDDCManager where needed.

File: playbooks/DeploySDDCMgr.yml [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+                                      SDDCMgr FIPS_Enable: {{ Nested_SDDCMgr.Component.Appliance.FIPSEnable }}
+
+                                       SDDCMgr Media Path: {{ Deploy.Software.SDDCMgr.Directory }}
+                                       SDDCMgr Media File: {{ Deploy.Software.SDDCMgr.File }}
+
+          =================================================================================================
+      when:
+        - DEBUG.DisplayVariables
+
+    - name: Check if SDDCMgr is already installed
```
Comment (rutgerblom) [inline playbooks/DeploySDDCMgr.yml:106]: SDDCManager

File: software_sample.yml [hunk 23]
```diff
@@ -1276,6 +1276,26 @@ Software:
           Version: "8.8"
           FileExt: iso
 
+##
```
Comment (rutgerblom) [inline software_sample.yml:1279]: SDDCMANAGER

File: templates/Pod_Config.j2 [hunk 24]
```diff
@@ -59,6 +59,8 @@ Deploy:
       Deploy: {{ Deploy.Product.DNSServer.Deploy }}
     Replication:
       Deploy: {{ Deploy.Product.Replication.Deploy }}
+    SDDCMgr:
+      Deploy: {{ Deploy.Product.SDDCMgr.Deploy }}
```
Comment (rutgerblom) [inline templates/Pod_Config.j2:63]: SDDCManager

File: templates/Pod_Config.j2 [hunk 25]
```diff
@@ -182,7 +184,16 @@ Deploy:
       Directory: "{{ Software[Deploy.Software.Replication.Vendor][Deploy.Software.Replication.Product]['Installers'][Deploy.Software.Replication.Version]['Location'].Local }}"
       File:      "{{ Software[Deploy.Software.Replication.Vendor][Deploy.Software.Replication.Product]['Installers'][Deploy.Software.Replication.Version].File }}"
       FileExt:   "{{ Software[Deploy.Software.Replication.Vendor][Deploy.Software.Replication.Product]['Installers'][Deploy.Software.Replication.Version].FileExt }}"
-
+    SDDCMgr:
```
Comment (rutgerblom) [inline templates/Pod_Config.j2:187]: SDDCManager

File: templates/Pod_Config.j2 [hunk 26]
```diff
@@ -213,6 +224,10 @@ Nested_Replication:
   {{ Nested_Replication | to_nice_yaml(indent=2, width=99999) | indent(2) }}
 
 
+Nested_SDDCMgr:
```
Comment (rutgerblom) [inline templates/Pod_Config.j2:227]: SDDCManager

File: playbooks/DeploySDDCMgr.yml
Comment (rutgerblom) [inline playbooks/DeploySDDCMgr.yml:1]: Replace SDDCMgr with SDDCManager

File: playbooks/CreatePodConfig.yml (hunk 18 above)
Comment (rutgerblom) [inline playbooks/CreatePodConfig.yml]: Change to _SDDCv

**Comment by rutgerblom** [comment]
> @rexit1982 feedback on two items:
> 
> 1. Type-O in comment of `DeploySDDCMgr.yml`.  `SDDG` should be `SDDC`.  Not a big deal, but since I'm entering this comment, thought I'd mention it.
> 2. I can foresee where someone might want to deploy SDDC Manager in a deployment regardless of the DeployESXiOnly setting.   If they set it to false, then all other products will deploy in the Pod, which can then be imported.  If they set it to true, then they want to use SDDC Manager to deploy all the products.  So, not sure you need the `Deploy.Setting.DeployESXiOnly == false` check on the 'when' statement for those reasons.
> 
> @rutgerblom , do you agree?
> 
> Luis

I agree with Luis. Let's go for **SDDCManager**. I started to comment on each line containing "SDDCMgr" but removed those comments again because they became too many. I think you know where to make the replacements :-)

File: playbooks/DeploySDDCMgr.yml [hunk 27]
```diff
@@ -0,0 +1,182 @@
+##
+##    Project: SDDC.Lab
+##    Authors: Luis Chanu & Rutger Blom
```
Comment (rutgerblom) [inline playbooks/DeploySDDCMgr.yml]: Authors: Aaron Ellis

**Comment by rexit1982** [comment]
> > @rexit1982 feedback on two items:
> > 
> > 1. Type-O in comment of `DeploySDDCMgr.yml`.  `SDDG` should be `SDDC`.  Not a big deal, but since I'm entering this comment, thought I'd mention it.
> > 2. I can foresee where someone might want to deploy SDDC Manager in a deployment regardless of the DeployESXiOnly setting.   If they set it to false, then all other products will deploy in the Pod, which can then be imported.  If they set it to true, then they want to use SDDC Manager to deploy all the products.  So, not sure you need the `Deploy.Setting.DeployESXiOnly == false` check on the 'when' statement for those reasons.
> > 
> > @rutgerblom , do you agree?
> > Luis
> 
> I agree with Luis. Let's go for **SDDCManager**. I started to comment on each line containing "SDDCMgr" but removed those comments again because they became too many. I think you know where to make the replacements :-)

Gotcha,  Made the changes/corrections running a test deployment before I push them.

File: playbooks/CreatePodConfig.yml (hunk 18 above)
Comment (rexit1982) [reply playbooks/CreatePodConfig.yml]: updated

File: playbooks/DeploySDDCMgr.yml (hunk 27 above)
Comment (rexit1982) [reply playbooks/DeploySDDCMgr.yml]: Changes made

**Comment by luischanu** [comment]
@rexit1982 thank you for making those changes, and those changes look good to me from my brief cursory review.

While reviewing, however, I did noticed a few other items that will ultimately need to get updated.  If you have the time to update these items, we would appreciate it:
1. Update the `UpdateDNS.yml` and `CleanupDNS.yml` playbooks so that DNS IPv4/IPv6 forward and reverse records get properly updated.
2. Update `Undeploy.yml` so that the SDDC Manager is "cleaned-up" when a Pod is destroyed.
3. @rutgerblom and I have talked about keeping the README.md file current as new features are added to the SDDC.Lab project.  With that in mind, if you could please add a `SDDC Manager (v8)` section at the end of the `Project Features` section and provide a brief summary of what this feature does, along with any instructions/insights/gotchas that you think would help the reader take advantage of this new feature, that would be helpful and greatly appreciated.

Thank you for taking the time to develop this PR to make the SDDC.Lab project even better!

Sincerely,

Luis Chanu

**Comment by rexit1982** [comment]
> @rexit1982 thank you for making those changes, and those changes look good to me from my brief cursory review.
> 
> While reviewing, however, I did noticed a few other items that will ultimately need to get updated. If you have the time to update these items, we would appreciate it:
> 
> 1. Update the `UpdateDNS.yml` and `CleanupDNS.yml` playbooks so that DNS IPv4/IPv6 forward and reverse records get properly updated.
> 2. Update `Undeploy.yml` so that the SDDC Manager is "cleaned-up" when a Pod is destroyed.
> 3. @rutgerblom and I have talked about keeping the README.md file current as new features are added to the SDDC.Lab project.  With that in mind, if you could please add a `SDDC Manager (v8)` section at the end of the `Project Features` section and provide a brief summary of what this feature does, along with any instructions/insights/gotchas that you think would help the reader take advantage of this new feature, that would be helpful and greatly appreciated.
> 
> Thank you for taking the time to develop this PR to make the SDDC.Lab project even better!
> 
> Sincerely,
> 
> Luis Chanu

Done

**Pull request merged**